test/e2e/node/runtimeclass.go

```diff
@@ -0,0 +1,190 @@
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/kubelet/events"
+	runtimeclasstest "k8s.io/kubernetes/pkg/kubelet/runtimeclass/testing"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	utilpointer "k8s.io/utils/pointer"
+
+	. "github.com/onsi/ginkgo"
+)
+
+const runtimeClassCRDName = "runtimeclasses.node.k8s.io"
+
+var (
+	runtimeClassGVR = schema.GroupVersionResource{
+		Group:    "node.k8s.io",
+		Version:  "v1alpha1",
+		Resource: "runtimeclasses",
+	}
+)
+
+var _ = SIGDescribe("RuntimeClass [Feature:RuntimeClass]", func() {
+	f := framework.NewDefaultFramework("runtimeclass")
+
+	It("should reject a Pod requesting a non-existent RuntimeClass", func() {
+		rcName := f.Namespace.Name + "-nonexistent"
+		pod := createRuntimeClassPod(f, rcName)
+		expectSandboxFailureEvent(f, pod, fmt.Sprintf("\"%s\" not found", rcName))
+	})
+
+	It("should run a Pod requesting a RuntimeClass with an empty handler", func() {
+		rcName := createRuntimeClass(f, "empty-handler", "")
+		pod := createRuntimeClassPod(f, rcName)
+		expectPodSuccess(f, pod)
+	})
+
+	It("should reject a Pod requesting a RuntimeClass with an unconfigured handler", func() {
+		handler := f.Namespace.Name + "-handler"
+		rcName := createRuntimeClass(f, "unconfigured-handler", handler)
+		pod := createRuntimeClassPod(f, rcName)
+		expectSandboxFailureEvent(f, pod, handler)
+	})
+
+	It("should reject a Pod requesting a deleted RuntimeClass", func() {
+		rcName := createRuntimeClass(f, "delete-me", "")
+
+		By("Deleting RuntimeClass "+rcName, func() {
+			err := f.DynamicClient.Resource(runtimeClassGVR).Delete(rcName, nil)
+			framework.ExpectNoError(err, "failed to delete RuntimeClass %s", rcName)
+
+			// Wait for the RuntimeClass to disappear.
+			framework.ExpectNoError(wait.PollImmediate(framework.Poll, time.Minute, func() (bool, error) {
+				_, err := f.DynamicClient.Resource(runtimeClassGVR).Get(rcName, metav1.GetOptions{})
+				if errors.IsNotFound(err) {
+					return true, nil // done
+				}
+				if err != nil {
+					return true, err // stop wait with error
+				}
+				return false, nil
+			}))
+		})
+
+		pod := createRuntimeClassPod(f, rcName)
+		expectSandboxFailureEvent(f, pod, fmt.Sprintf("\"%s\" not found", rcName))
+	})
+
+	It("should recover when the RuntimeClass CRD is deleted [Slow]", func() {
+		By("Deleting the RuntimeClass CRD", func() {
+			crds := f.APIExtensionsClientSet.ApiextensionsV1beta1().CustomResourceDefinitions()
+			runtimeClassCRD, err := crds.Get(runtimeClassCRDName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get RuntimeClass CRD %s", runtimeClassCRDName)
+			runtimeClassCRDUID := runtimeClassCRD.GetUID()
+
+			err = crds.Delete(runtimeClassCRDName, nil)
+			framework.ExpectNoError(err, "failed to delete RuntimeClass CRD %s", runtimeClassCRDName)
+
+			// Wait for the CRD to disappear.
```

Comment:
nit: usually prefer using `By` when annotation sections of code like this.

Reply:
Done.